feat: add navigation confirmation when builder fields are dirty

karrui opened the pull request:
## Problem
<!-- What problem are you trying to solve? What issue does this close? -->
This PR adds a navigation confirmation on navigation and change of field when the field being currently edited is dirty. The possible navigation cases are covered below:

- start page
- end page
- field
- builder tabs
- drawer close button
- drawer back button

Any permutation (start page -> field, field -> field, end page -> start page) will trigger a render of a new `DirtyModal` component if the fields are dirty. This was achieved by storing pending states in the respective stores and context, subscribing to a newly exposed `isDirty` selector in `fieldBuilderStore`. 

Any route navigation change will also trigger a similar `NavigationPrompt` modal.

Closes #4304

## Solution
<!-- How did you solve the problem? -->

**Breaking Changes** 
<!-- Does this PR contain any backward incompatible changes? If so, what are they and should there be special considerations for release? -->
- [x] No - this PR is backwards compatible  

**Features**:

- feat: add `NavigationPrompt` modal component to show navigation confirmation modal when condition is true
  - this is used in this PR when the `isDirty` constant is true.
- feat: update `useDesignStore`, `useFieldBuilderStore`, `CreatePageSidebarContext` to be responsive to dirty field state. 
- feat: show confirmation modal when navigating or changing edit fields when current field is dirty.

## Screenshots
CreatePage stories should now have the behaviour, but here is a gif in case that is hard to navigate.
![Recording 2022-08-25 at 12 06 00](https://user-images.githubusercontent.com/22133008/186572459-13ef69b1-8b9b-40fd-8de2-671276c761af.gif)

karrui committed:
wip: pass isdirty state to builder store for actioning

karrui committed:
wip: add NavigationPrompt modal component to show for nav confirmation

karrui committed:
feat(LandingPage): remove commented code

karrui committed:
feat: trigger NavigationPrompt on dirty field and route navigation

karrui committed:
ref: extract logic to useNavigationPrompt hook

karrui committed:
fix(useNavigationPrompt): remove unnecessary window prompt on load

karrui committed:
feat: add navigation blocker to start and end page edits

karrui committed:
feat: show modal if clicking into other field on current dirty drawer

karrui committed:
feat: update design store to store holding state for navigation

karrui committed:
feat: update dirty modal to handle start page de-navigation

karrui committed:
feat: handle admin builder dirty endpage click

karrui committed:
feat: add dirtymodal handler when changing builder tabs

karrui committed:
feat: add pending state for clicking builder tabs

karrui committed:
feat: prevent reordering when fields are dirty

karrui committed:
fix: add correct invocation of closeBuilderDrawer

justynoh commented:
Regarding the UX itself, is there any particular reason to have different modals in each case? Since there are two sets of modals (one for navigation and one for changing the active state within the same route), I wonder if users would have the (incorrect) impression that "discarding changes" means different things for each modal. 

For instance (and this is just my intuition - feel free to comment if you don't think so), if i click out of a dirty field, I might think that the "You have unsaved changes" modal will discard my changes to the current active field. But if let's say i've been working on my form and I try to navigate away, seeing a different modal ("Discard changes?") I might be worried that there are other changes to the form that aren't yet saved.

karrui commented:
> Regarding the UX itself, is there any particular reason to have different modals in each case? Since there are two sets of modals (one for navigation and one for changing the active state within the same route), I wonder if users would have the (incorrect) impression that "discarding changes" means different things for each modal.
> 
> For instance (and this is just my intuition - feel free to comment if you don't think so), if i click out of a dirty field, I might think that the "You have unsaved changes" modal will discard my changes to the current active field. But if let's say i've been working on my form and I try to navigate away, seeing a different modal ("Discard changes?") I might be worried that there are other changes to the form that aren't yet saved.

@pearlyong?

justynoh commented on frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignContent/BuilderAndDesignContent.tsx:
```diff
@@ -36,7 +36,9 @@ export const BuilderAndDesignContent = ({
-  useEffect(() => setFieldsToInactive, [setFieldsToInactive])
+  useEffect(() => {
+    setFieldsToInactive()
+  }, [setFieldsToInactive])
```
Should we also `setFieldsToInactive` on unmount like we did previously so that we keep the state clean when the admin exits the builder?

justynoh commented on frontend/src/features/admin-form/create/builder-and-design/useFieldBuilderStore.tsx:
```diff
@@ -11,64 +11,116 @@ export enum FieldBuilderState {
+  // Will be used to set stateData if user confirms discarding changes.
+  holdingStateData:
+    | FieldBuilderCreateEditStateData
+    | { state: FieldBuilderState.Inactive }
```
what's the reason for keeping the inactive state separate from `FieldBuilderCreateEditStateData`? didn't seem like there's anywhere that requires them to be separate

karrui replied:
reusing `FieldBuilderCreateEditStateData` type elsewhere where we want to ensure inactive state cannot be set.

justynoh commented on frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/EditFieldDrawer/edit-fieldtype/common/useEditFieldForm.ts:
```diff
@@ -65,18 +67,25 @@ export const useEditFieldForm = <FormShape, FieldShape extends FormField>({
-        [],
-      ),
-    )
+  const {
+    stateData,
+    setToInactive,
+    updateEditState,
+    updateCreateState,
+    setIsDirty,
+  } = useFieldBuilderStore(
+    useCallback(
+      (state) => ({
+        stateData: stateDataSelector(state),
+        setToInactive: setToInactiveSelector(state),
+        updateEditState: updateEditStateSelector(state),
+        updateCreateState: updateCreateStateSelector(state),
+        isDirty: isDirtySelector(state),
+        setIsDirty: setIsDirtySelector(state),
+      }),
+      [],
+    ),
+  )
```
seems like `isDirty` isn't included in the destructuring here, can we take that line out?

justynoh commented on frontend/src/templates/NavigationPrompt/useNavigationPrompt.ts:
```diff
@@ -0,0 +1,67 @@
+  ) as NavigationContextWithBlock
+
+  const [isPromptShown, setIsPromptShown] = useState(false)
+  const [currentPath, setCurrentPath] = useState('')
```
is `targetPath` a better name for this state?

justynoh commented on frontend/src/features/admin-form/create/builder-and-design/useDesignStore.tsx:
```diff
@@ -9,7 +9,7 @@ import { UploadedImage } from './BuilderAndDesignDrawer/EditFieldDrawer/edit-fie
 
 export enum DesignState {
-  EditingHeader,
+  EditingHeader = 1,
```
why is this `=1` needed?

karrui replied:
We want the enum to start from `1` (so falsey footguns, like when value is `0` and you do a negate check is less likely to happen)

justynoh commented on frontend/src/features/admin-form/create/builder-and-design/useFieldBuilderStore.tsx:
```diff
@@ -11,64 +11,116 @@ export enum FieldBuilderState {
+  updateEditState: (field: FormFieldDto, holding?: boolean) => void
+  setEditEndPage: (holding?: boolean) => void
+  setToInactive: (holding?: boolean) => void
+  isDirty: boolean
+  setIsDirty: (isDirty: boolean) => void
```
since these are used throughout design and endpage too, would it be better to put them somewhere above, perhaps in the Create page context or Admin builder context? Because the original intention for separating the design, builder and endpage stores was to separate the concerns, thus putting `isDirty` here feels a bit counter to that as its also used in design and endpage

karrui replied:
didn't want to put something that changes so often into context, since that will trigger rerenders. Opted to put in this store first, we can always move it in the future. Seemed overkill to add yet another store just to hold one dirty state. do you think we should create a new store? actually ok let me do it

karrui replied:
done in 4ecda0b

karrui committed:
feat(BuilderContent): setFieldsToInactive on unmount too

karrui committed:
feat(useEditFieldForm): remove unused store selector

karrui committed:
feat(useNavigationPrompt): rename to targetPath

karrui committed:
feat: add useDirtyFieldStore zustand store for dirty state tracking

karrui committed:
Merge branch 'form-v2/develop' into form-v2/fix-issue-4304-confirm-navigation

# Conflicts:
#	frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignContent/FieldRow/FieldRowContainer.tsx

pearlyong commented:
agree it might be confusing, let's stick with the 'You have unsaved changes' modal, which is easier English to understand

karrui committed:
fix: bug where clicking on placeholder would trigger navigation prompt

karrui committed:
feat: add explicit pending boolean parameter

removes the possibility of passing the function directly into a on* handler which causes the pending argument to be truthy

karrui committed:
ref: extract and use common UnsavedChangesModal

karrui committed:
fix: prevent save modal from showing when editing diff startpage field

justynoh reviewed (approved):
lgtm
